massa-final-state/src/final_state.rs

```diff
@@ -235,12 +189,96 @@ impl FinalState {
             .expect("Cycle history should not be empty in snapshot!")
             .clone();
 
-        // Reset the cycle_history and populate it with new cycle
-        self.pos_state.cycle_history = Vec::new().into();
-        self.pos_state
-            .create_new_cycle_for_snapshot(&latest_consistent_cycle_info, end_slot);
+        if current_slot_cycle == end_slot_cycle {
+            // In that case, we just complete the gap in the same cycle
+            let latest_cycle_info =
+                self.pos_state
+                    .cycle_history
+                    .back_mut()
+                    .ok_or(FinalStateError::SnapshotError(String::from(
+                        "Invalid cycle_history",
+                    )))?;
+
+            for _ in current_slot.period..end_slot.period {
+                latest_cycle_info.rng_seed.push(false);
+            }
+        } else {
+            // Here, we we also complete the cycle_infos in between
+
+            // Firstly, complete the first cycle
+            let latest_cycle_info =
+                self.pos_state
+                    .cycle_history
+                    .back_mut()
+                    .ok_or(FinalStateError::SnapshotError(String::from(
+                        "Invalid cycle_history",
+                    )))?;
+
+            for _ in current_slot.period..self.config.periods_per_cycle {
+                latest_cycle_info.rng_seed.push(false);
+            }
+            latest_cycle_info.complete = true;
+
+            // feed final_state_hash to the completed cycle
+            self.pos_state
+                .feed_cycle_state_hash(current_slot_cycle, self.final_state_hash);
+
+            self.pos_state
+                .feed_selector(current_slot_cycle.checked_add(2).ok_or_else(|| {
+                    FinalStateError::PosError("cycle overflow when feeding selector".into())
+                })?)
+                .map_err(|_| {
+                    FinalStateError::PosError("cycle overflow when feeding selector".into())
+                })?;
+
+            // Then, build all the already completed cycles
+            for cycle in current_slot_cycle + 1..end_slot_cycle {
+                // TODO: build a new cycle by repeating the one before. How do we handle this if latest cycle info is not complete?
+                let last_slot = Slot::new_last_of_cycle(
+                    cycle,
+                    self.config.periods_per_cycle,
+                    self.config.thread_count,
+                )
+                .map_err(|err| {
+                    FinalStateError::InvalidSlot(format!(
+                        "Cannot create slot for interpolating downtime: {}",
+                        err
+                    ))
+                })?;
```

Comment:
the `last_slot` creation inside the for loop can be done in the iterator creation itself. This suggested change (not to be commited directly) forms a template such that the body of this loop is strictly about behavior, without any value declaration distractions. This should allow better readability for future maintainers>

```suggestion
            // We now need to visit all the slots in the already completed cycles... 
            let completed_slots_iter = (current_slot_cycle + 1..end_slot_cycle).map(|cycle| Slot::new_last_of_cycle(cycle, ...).map_err(|err| {
                    FinalStateError::InvalidSlot(format!(
                        "Cannot create slot for interpolating downtime: {}",
                        err
                    ))
                });
            // ...and at each slot, we need to by $404-reason-not-found
            for last_lot in slot_iter {
            
```

Reply:
I'm not sure about this one, I'll try to refactor to make it clearer (maybe with helper functions and comments), but IMO it's clearer to thing about: 
> "I need to create the cycle N, and for that I need info on the last_slot of cycle N"

than to think about

> "I'll extend the cycle_info to cover everything until last_slot, that conveniently is the last slot of a cycle I'm trying to build"

Reply:
Ideally, you could do something like this

`(....).try_map($MAKE_SLOT)?.for_each($DO_POS_STATE_THINGS)`

...bat that's non-trivial without doing the whole construction first.

If this is a common loop, and pos-state activities are expensive, it might be worth considering building all the slots first anyway, to make sure it's a valid exercise. The idea being is that you don't want to get to the very end, having done all that, and fail.

...but it's a discression thing. This isn't worth spending too much time or energy on.

Reply:
As it's ran only once, I'll let it as is right now. However, I keep the note for the follow-up next month ;)